### PR TITLE
add some examples to the stdlib's documentation

### PR DESCRIPTION
**Pull request by c-cube**

### CODE REVIEW

File: stdlib/format.mli
Comment (bluddy) [inline stdlib/format.mli]: Could we use a shorter example? Like to 10 or 15?

Comment (c-cube) [reply stdlib/format.mli]: trying with 20, should be enough

File: stdlib/atomic.mli
Comment (gasche) [inline stdlib/atomic.mli]: `;;` is not idiomatic OCaml -- and not necessary for toplevel copyability either. Can you explain why you want to keep them around?

File: stdlib/atomic.mli
Comment (gasche) [inline stdlib/atomic.mli]: I find it a bit awkward to exemplify concurrent atomics with a single-domain example that does not actually need atomics. Alternatives include using `Domain` directly (con: arguably too low-level in general) or domainslib (con: outside the stdlib, may change or get replaced in the medium-term future).

File: stdlib/format.mli
Comment (gasche) [inline stdlib/format.mli]: Nitpick: this paragraph uses impersonal "one" which is a bit sophisticated, and "it's" at the same time which slightly more familiar than our usual documentation. I would rewrite it to smooth it out by avoiding both forms.

> It is good practice to define formatters for important types of your program.
> For example, consider basic geometry types:

File: stdlib/hashtbl.mli
Comment (gasche) [inline stdlib/hashtbl.mli]: `{b Warning}` is used elsewhere

File: stdlib/queue.mli
Comment (gasche) [inline stdlib/queue.mli]: I'm surprised by the amount of empty lines in this example, it rather makes the code harder to read for me, but it's okay if it's your style / you have a strong preference.

File: stdlib/queue.mli
Comment (gasche) [inline stdlib/queue.mli]: I would check `explored` before calling `p`, as the latter may be arbitrarily expensive so we don't want to call it several times on the same node.

**Review by gasche — COMMENTED** [review]
Thanks! I think that this is an excellent idea. See inline comments.

**Comment by shindere** [comment]
@pierreweis may be interested in reviewing this.

File: stdlib/queue.mli
Comment (c-cube) [reply stdlib/queue.mli]: I could add comments. I've tended to add a few empty lines to separate the phases, but it might just be better in a more compact style indeed.

File: stdlib/atomic.mli
Comment (c-cube) [reply stdlib/atomic.mli]: I've been testing it in a 4.14 repl on the side (in which it was surprisingly hard to get the `Thread` toplevel module?!), so I don't know exactly how I should go about that. Maybe I just need a 5.0-alpha switch to run this.

`Domain.spawn` should be on par in terms of complexity with `Thread.create`, right? Ignoring the ticker thread…

File: stdlib/atomic.mli
Comment (c-cube) [reply stdlib/atomic.mli]: They are useful if  you send phrases to the toplevel one by one. I can remove them.

Comment (gasche) [reply stdlib/atomic.mli]: Unless you have a strong opinion, I would remove them. I think it's better to show the code as we would actually write it in a source file. (Note: Tuareg has a shortcut to send a sentence to the toplevel, which adds `;;` automatically, doesn't your editor support something like this?)

File: stdlib/atomic.mli
Comment (gasche) [reply stdlib/atomic.mli]: Another option is to show not a complete program with Domain.spawn, but just the code of a library function that *may* be called concurrently. (For example: a "unique id" generator that works by incrementing an internal counter.) You can show the sequential version, point out that it is not domain-safe, and move to Atomic instead.

Comment (c-cube) [reply stdlib/atomic.mli:53]: removed

Comment (c-cube) [reply stdlib/atomic.mli]: moved to `Domain`. Hopefully someone can check the code, unless I get to trying 5.0-alpha on that.

**Comment by dbuenzli** [comment]
Could we please have the examples at the end of the modules in a dedicated section (e.g. `{1:examples Examples}}`) rather than in the module preamble (where a simple `See {{!examples}examples}.}` will do).

If you write them in the module preamble they get in the way when you peruse the API reference which what you eventually end up doing 99.9% of the time.

File: stdlib/hashtbl.mli
Comment (ulugbekna) [inline stdlib/hashtbl.mli]: There seems to be some repetition.

File: stdlib/atomic.mli
Comment (ulugbekna) [inline stdlib/atomic.mli]: Wouldn't an even simpler example work? e.g., `n` domains each incrementing the counter k times by 1

I am just inclined to think that one shouldn't really need to understand how to work with files to understand how to use atomics.

File: stdlib/hashtbl.mli
Comment (ulugbekna) [inline stdlib/hashtbl.mli]: This statement seems to me a little unfriendly for beginners. 

Maybe something in the line "Since most operations over this data structure are in-place, i.e., mutating, it is more common in imperative code." would work better? 

Also, while I understand what you mean by "fast lookup", it a little ambiguous.

File: stdlib/queue.mli
Comment (ulugbekna) [inline stdlib/queue.mli]: BFS through a tree would be less involved?

**Review by ulugbekna — COMMENTED** [review]
Thanks for the PR! I left a couple of nits if that's okay :-)

File: stdlib/atomic.mli
Comment (c-cube) [reply stdlib/atomic.mli]: I like to have "real" examples, I suppose. What do the maintainers think? @Octachron @gasche for example

File: stdlib/queue.mli
Comment (c-cube) [reply stdlib/queue.mli]: I've never really needed a BFS through a tree. I think BFS through a graph (of some kind) is one of the main application of queues, so I illustrate with a real example. Same as above I suppose, it depends on what maintainers think the docs of the stdlib should look like.

File: stdlib/atomic.mli
Comment (dbuenzli) [inline stdlib/atomic.mli]: Not sure why you get down to `{3` here, these sections should be `{2`. It would also be nice to label them. So that they can be hyperlinked via for example `{!Atomic.basic_counter}`. 

Same in `queue.mli`

File: stdlib/format.mli
Comment (dbuenzli) [inline stdlib/format.mli]: We know we are on the `Format` page. `Examples` is fine in my opinion (and more consistent).

File: stdlib/atomic.mli
Comment (gasche) [reply stdlib/atomic.mli]: I found this example a tad long as well, on the other hand the BFS sounded like a natural choice for queues. (I'm often too lazy to reach for a queue when I BFS and I use two lists, "frontier" and "next", but let's forget about this nitpick.) The Atomic example also had the issue of using Thread instead of Domain. It's fixed now (I haven't tried to run it yet, I'm in holidays until the end of the month and I somehow convinced myself that replying to messages is okay but running the compiler on PRs is not), but there is still a wide margin for debate. For example, the message we send is that `Domain` is the low-level, core module and that users should use higher-level abstractions (outside for the stdlib for now).

I think that the duty is on us to find a "real example" that is simpler but that @c-cube still likes, and maybe also manages to tiptoe around those difficulties.

**Comment by gasche** [comment]
> What do the maintainers think?

Meta point: I think that overall the PR is an excellent idea and we should try to converge by making as many people as possible happy about it. I'm happy to play tie-breaker if it is productive and works in practice, but overall let's just roll with the comments until they quiet down and merge the result?

**Comment by nojb** [comment]
I also think it is an excellent idea to add examples and tutorials to the Stdlib. On Format in particular, I always found the old https://caml.inria.fr/resources/doc/guides/format.en.html quite useful; perhaps it can serve as inspiration...

File: stdlib/atomic.mli
Comment (c-cube) [reply stdlib/atomic.mli]: fixed

**Comment by c-cube** [comment]
@nojb I've never been too fond of this tutorial :slightly_smiling_face:. I think it can be included/linked to in another PR anyway. One thing at a time!

My (obviously) objective and unbiased view is that the examples I added here are more straight to the point (but don't explain as much). There's no box theory at all, just examples that are (hopefully) detailed.

**Comment by xavierleroy** [comment]
Like @gasche I'm very happy to see a concrete proposal to add examples to the stdlib.

I was about to +1 @dbuenzli's suggestion of putting the examples at the end of the .mli file, but it seems done already, thanks!

Of course we need to discuss the examples and the explanations :-)  For example I find the first example for Atomic too long-winded; I think it would suffice to just show two domains that do `Atomic.incr` on the same atomic integer, and contrast with two domains that do plain `incr`.

Also, now would be a great time to reorder the declarations in .mli files and group them logically into sections, so that the simpler and more generally useful functions come first.  @dbuenzli did this for the String module, if I remember correctly, but Hashtbl and other modules would benefit too.

**Comment by dbuenzli** [comment]
> other modules would benefit too.

For `Set` and `Map` it's waiting for reviews in https://github.com/ocaml/ocaml/pull/11410

**Comment by xavierleroy** [comment]
> For Set and Map it's waiting for reviews in https://github.com/ocaml/ocaml/pull/11410

Oh, I missed this one.  Thanks for reminding me, and keep up the good work :-)

**Comment by gasche** [comment]
What do we need to move this PR forward?

I think that "restructuring the .mli file" could be split off to a separate PR. My understanding is that people would like a better/different example for the atomics, and are basically fine with the rest.

**Comment by c-cube** [comment]
I'm not sure what the simpler example could be. I agree that the IO stuff is a bit too long (even though it is, imho, a real use case for atomic counters: metrics).

**Comment by Octachron** [comment]
Concerning the examples in `Atomic`, I have the impression that it might work better to drop the first example and expand the second example to explain that the use of atomic guarantees that not elements are silently dropped from the queue.

**Comment by c-cube** [comment]
I expanded a bit the second example, and replaced the first one with a groundbreaking Proof Of Work™  implementation utilizing multiple cores to break Difficult Mathematical Problems™.

Basically just showing how to use `bool Atomic.t` to stop multiple threads; and a global thread-safe iteration counter. Both are best expressed with atomic.

File: stdlib/atomic.mli
Comment (gasche) [inline stdlib/atomic.mli:101]: Instead of `8` you could use `Domain.recommended_domain_count` here. (This helps drive down the point that domains are not lightweight threads you can spawn as many as you want.)

File: stdlib/atomic.mli
Comment (gasche) [inline stdlib/atomic.mli:84]: Now that I became aware of splitting random generators, I would use a different approach where I seed a generator once at toplevel, and then split it repeatedly (on the main domain), passing a split state to `find_number_where`. This lets you easily control seeding (non-deterministic or deterministic are both possible). But the current approach is slightly simpler and also fine, this example is not about Random itself.

**Review by gasche — APPROVED** [review]
I like the new first example of Atomic better, and overall I like the PR.

I think we should consider merging it; it's better to have examples than not, and I'm happy with the current ones. (We can always refine a bit later if we have concerns.)

Approved. (Note that stdilb PRs require two approvals.)

File: stdlib/format.mli
Comment (nojb) [inline stdlib/format.mli]:
```suggestion
        (6, true); (7, false); (8, true); (9, false); (10, true); (11, false);
        (12, true); (13, false); (14, true); (15, false); (16, true);
        (17, false); (18, true); (19, false)]
```

File: stdlib/format.mli
Comment (nojb) [inline stdlib/format.mli]:
```suggestion
      "@[<hov>" and "@]" is a "horizontal-or-vertical box".
```
The rule as far as I know it is that `an` should be used when the following _sound_ is a vowel sound, but the `h` in horizontal is not silent, so `a` should be used.

File: stdlib/format.mli
Comment (nojb) [inline stdlib/format.mli]:
```suggestion
  For debugging purposes, or to display information
  in logs or the console, it would be convenient to define printers for these types. Here is an example of how to do
  it. Note that "%.3f" is a [float] printer with 3 digits of precision
  after the dot; "%f" would print as many digits as required, which is
  somewhat verbose; "%h" is a hexadecimal float printer.
```

File: stdlib/format.mli
Comment (nojb) [inline stdlib/format.mli]:
```suggestion
  A final note: the [Format] module is a starting point.
```

File: stdlib/hashtbl.mli
Comment (nojb) [inline stdlib/hashtbl.mli]:
```suggestion
    let seq = Seq.ints 0 |> Seq.take 100
```

File: stdlib/hashtbl.mli
Comment (nojb) [inline stdlib/hashtbl.mli:576]:
```suggestion
  Here we illustrate that principle using a sequence of (ASCII) characters
```

File: stdlib/moreLabels.mli
Comment (nojb) [inline stdlib/moreLabels.mli]:
```suggestion
      let seq = Seq.ints 0 |> Seq.take 100
```

File: stdlib/queue.mli
Comment (nojb) [inline stdlib/queue.mli]:
```suggestion
      in
      loop()
```

File: stdlib/templates/hashtbl.template.mli
Comment (nojb) [inline stdlib/templates/hashtbl.template.mli]:
```suggestion
    let seq = Seq.ints 0 |> Seq.take 100
```

**Review by nojb — APPROVED** [review]
LGTM (modulo a few minor stylistic comments)

File: Changes
Comment (gasche) [inline Changes]: Could you complete the list of reviewers? (There is a "reviewers" list in the top right of the PR main webpage, which is accurate as far as I can tell.)

**Comment by nojb** [comment]
Merged, thanks!

**Comment by c-cube** [comment]
Thank  you @nojb and to all the reviewers! :grin:

**Pull request merged**
